### PR TITLE
Range liquidity

### PR DESCRIPTION
**Pull request by JasonMHasperhoven**
Implements https://github.com/orgs/prax-wallet/projects/1/views/7?pane=issue&itemId=87984824&issue=prax-wallet%7Cprax%7C239

### CODE REVIEW

**Comment by JasonMHasperhoven** [comment]
I need some help, something is wrong in the logic here:

image

**Comment by hdevalence** [comment]
Awesome progress! Some comments from a trial run:

Screenshot 2024-11-29 at 3 06 48 AM

This looks wrong -- the "liquidity provided" at the top should be the total amount of liquidity, priced in terms of asset 2 at the current price, i.e., we are aiming to have the present value of all of the newly created positions sum up to the target. But here, it looks like we get 20 USDC and ~20 UM, approximately double the exposure that's desired. To do this correctly, I think the move is:

1. Take the total target amount (in asset 2)
2. Divide by the number of positions to get the per-position liquidity amount (in asset 2)
3. Compute the equivalent amount of asset 1 (at current market price)
4. For each position, if the position's price is less than the market price, initialize with per-position amonut of asset 1, otherwise with per-position amount of asset 2

Then the total value of all the positions should sum to the target amount.

Screenshot 2024-11-29 at 3 06 14 AM

Looks like this is overflowing the height of the container. Also not sure why there's black text that's mirroring the text inputs? Also having the issue where I can't backspace the "0" in the fee selector box, it only allows me to type "04" and then to delete the leading 0.

### Nonblocking / Tangents

Screenshot 2024-11-29 at 3 07 17 AM

We should figure out a way to extend the value view to use the extended_metadata field to render an LP position so these aren't "Unknown asset"s. This needs some design work + Prax plumbing (need to have a way to insert LPs into rendering of approval dialog when the LPs haven't been created yet), cc @TalDerei + Sam (blanking on sam's GH handle)

Screenshot 2024-11-29 at 3 07 38 AM

This popup isn't actionable. We should plan to integrate a transaction viewer into the Inspect tab so users can close the loop on their actions without leaving the app. Hopefully this can just pull in the V2 transaction view components.

**Comment by hdevalence** [comment]
Screenshot 2024-11-29 at 3 25 58 AM

I checked with `pcli` and it looks like the fee is not being set correctly - i set a 4% fee but it's created with 0

**Comment by hdevalence** [comment]
Screenshot 2024-11-29 at 3 26 37 AM

Also not sure if related to the PR but i get a bunch of error text filling the page (black on black - screenshot is highlighted) when i run the branch

**Comment by TalDerei** [comment]
> We should figure out a way to extend the value view to use the extended_metadata field to render an LP position so these aren't "Unknown asset"s. This needs some design work + Prax plumbing (need to have a way to insert LPs into rendering of approval dialog when the LPs haven't been created yet), cc @TalDerei + Sam (blanking on sam's GH handle)

tracked in https://github.com/prax-wallet/prax/issues/244. 

sam's twitter handle: https://x.com/dismissforever

> This popup isn't actionable. We should plan to integrate a transaction viewer into the Inspect tab so users can close the loop on their actions without leaving the app. Hopefully this can just pull in the V2 transaction view components.

captured by feedback in https://github.com/penumbra-zone/dex-explorer/issues/140

**Comment by JasonMHasperhoven** [comment]
@hdevalence The overflow issues is addressed.

The amount looks to me to be correct:

I'm checking the pair UM/GM with price 0.853844. Then to verify we can do 11.68829*0.853844=9.9799762868, which seems correct as 10. To get this amount we actually send a simulate trade request like below:

```
const req = new SimulateTradeRequest({
   input: assetIn.toValue(),
   output: assetOut.assetId,
});
```

image

**Comment by JasonMHasperhoven** [comment]
Next PR would be a limit order as individual position.

File: src/pages/trade/ui/order-form/store/index.ts [hunk 1]
```diff
@@ -0,0 +1,502 @@
...
+    this.quoteAsset = new OrderFormAsset(quoteAsset);
+
+    const debouncedHandleAmountChange = debounce(this.handleAmountChange, 500);
+
+    this.baseAsset.onAmountChange(
+      debouncedHandleAmountChange as (asset: OrderFormAsset) => Promise<void>,
+    );
+    this.quoteAsset.onAmountChange(
+      debouncedHandleAmountChange as (asset: OrderFormAsset) => Promise<void>,
+    );
```
[the remark this answers is not in the document]
Comment (grod220) [reply src/pages/trade/ui/order-form/store/index.ts:142]: Though, should it be cast? It seems like that debounce function can return undefined.

File: src/pages/trade/ui/order-form/store/asset.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  };
+
+  setAccountAddress = (addressView: AddressView): void => {
+    this.accountAddress = getAddress(addressView);
+    this.accountIndex = getAddressIndex(addressView);
+  };
+
+  setBalanceFromBalanceView = (balanceView: ValueView): void => {
+    const balance = getFormattedAmtFromValueView(balanceView, true);
+    this.balance = parseFloat(balance.replace(/,/g, ''));
```
[the remark this answers is not in the document]
Comment (JasonMHasperhoven) [reply src/pages/trade/ui/order-form/store/asset.ts:54]: I see what you mean. I can do this as part of the next PR if you don't mind?

**Pull request merged**